internal/node/fishtail/ingest.go

```diff
@@ -0,0 +1,229 @@
+	"github.com/bluesky-social/indigo/atproto/syntax"
+	"github.com/eagraf/habitat-new/internal/node/controller"
+	"github.com/rs/zerolog/log"
+
+	types "github.com/eagraf/habitat-new/core/api"
+)
+
+type Ingester struct {
+	// Queue of chains of records to ingest
+	chainQueue chan *RecordChainIngester
+	pdsClient  *controller.PDSClient
+}
+
+func NewIngester(pdsClient *controller.PDSClient) *Ingester {
+	return &Ingester{
+		chainQueue: make(chan *RecordChainIngester, 1000),
+		pdsClient:  pdsClient,
+	}
+}
+
+func (i *Ingester) EnqueueChain(chain *RecordChainIngester) {
+	chain.pdsClient = i.pdsClient
+	i.chainQueue <- chain
+}
+
+func (i *Ingester) Run() error {
+	// Right now, we just ingest one chain at a time from the queue.
+	for {
+		chain := <-i.chainQueue
+		log.Info().Msgf("got chain to ingest")
+
+		// Ingest the initial record, which is a special case because it comes as CBOR
+		// directly from the event stream.
+		err := chain.StartIngestion()
+		if err != nil {
+			log.Error().Msgf("failed to ingest initial record: %s", err)
+			continue
+		}
+		log.Info().Msgf("ingested initial record of collection %s", syntax.ATURI(chain.initialURI).Collection().String())
+
+		cont := true
+		for cont {
+			shouldContinue, err := chain.IngestNext()
+			if err != nil {
+				log.Error().Msgf("failed to ingest next record: %s", err)
+			}
+			cont = shouldContinue
+		}
+
+		err = chain.FinishIngestion()
+		if err != nil {
+			log.Error().Msgf("failed to finish ingestion: %s", err)
+		}
+	}
+}
+
+type RecordChainIngester struct {
+	initialCBORRecord []byte
+	initialCID        string
+	initialURI        string
+
+	toIngest  chan syntax.ATURI
+	pdsClient *controller.PDSClient
+
+	seen map[string]interface{}
+
+	ingestedRecords []*types.PDSGetRecordResponse
+
+	atProtoEventPublisher *ATProtoEventPublisher
+}
+
+type IngestedRecordChain struct {
+	Collection       string                        `json:"collection"`
+	InitialRecordURI string                        `json:"initial_record_uri"`
+	Records          []*types.PDSGetRecordResponse `json:"records"`
+}
+
+func NewRecordChainIngestor(
+	initialCBORRecord []byte,
+	initialCID string,
+	initialURI string,
+	atProtoEventPublisher *ATProtoEventPublisher,
+) *RecordChainIngester {
+	return &RecordChainIngester{
+		initialCBORRecord: initialCBORRecord,
+		initialCID:        initialCID,
+		initialURI:        initialURI,
+
+		toIngest:              make(chan syntax.ATURI, 1000),
+		seen:                  make(map[string]interface{}),
+		atProtoEventPublisher: atProtoEventPublisher,
+	}
+}
+
+func (ic *RecordChainIngester) EnqueueRecord(uri syntax.ATURI) bool {
+	log.Info().Msgf("enqueueing record: %s", uri)
+	if _, ok := ic.seen[uri.String()]; !ok {
+		ic.seen[uri.String()] = true
+		ic.toIngest <- uri
+		return true
+	}
+	return false
+}
+
+// IngestNext ingests the next record in the chain. If there are no more records to ingest, it returns false.
+func (ic *RecordChainIngester) IngestNext() (bool, error) {
```

Comment:
I'm a little confused about what the "Chain" is here? Are there docs you can link to explaining why we need to do this?